src/main.rs

```diff
@@ -24,50 +25,59 @@ use std::{
     fs::File,
     io::{Read, Seek, SeekFrom},
 };
+// use std::collections::HashMap;
+// type AHashMap<K, V> = HashMap<K, V>;
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Copy)]
 struct WeatherData {
-    total_temperature: f64,
-    min_temperature: f64,
-    max_temperature: f64,
+    total_temperature: f32,
+    min_temperature: f32,
+    max_temperature: f32,
     count: u32,
-    mean_temperature: f64,
+    mean_temperature: f32,
 }
 
 impl WeatherData {
+    #[inline(always)]
     fn merge(&mut self, other: &WeatherData) {
         self.total_temperature += other.total_temperature;
         self.count += other.count;
         self.min_temperature = self.min_temperature.min(other.min_temperature);
         self.max_temperature = self.max_temperature.max(other.max_temperature);
     }
-
-    fn add_temperature(&mut self, temperature: f64) {
+    #[inline(always)]
+    fn add_temperature(&mut self, temperature: f32) {
         self.min_temperature = self.min_temperature.min(temperature);
         self.max_temperature = self.max_temperature.max(temperature);
         self.total_temperature += temperature;
         self.count += 1;
     }
-
+    #[inline(always)]
     fn update_mean(&mut self) {
-        self.mean_temperature = self.total_temperature / self.count as f64;
+        self.mean_temperature = self.total_temperature / self.count as f32;
     }
-
+    #[inline(always)]
     fn round(&mut self) {
         self.mean_temperature = (self.mean_temperature * 10.0).round() / 10.0;
         self.min_temperature = (self.min_temperature * 10.0).round() / 10.0;
         self.max_temperature = (self.max_temperature * 10.0).round() / 10.0;
     }
 }
 
-fn process_weather_line(line: &str) -> Result<(&str, WeatherData), &'static str> {
+const KEY_SIZE: usize = 16;
+type Key = [u8; KEY_SIZE];
+type StationTemperatures = AHashMap<Key, WeatherData>;
```

Comment:
Everything looks perfect! But I have a comment here. I see the problem statement there can be 100 characters maximum in the station name field.
```
Station name: non null UTF-8 string of min length 1 character and max length 100 bytes,
    containing neither ; nor \n characters. (i.e. this could be 100 one-byte characters,
    or 50 two-byte characters, etc.)
```

So we can make this `KEY_SIZE = 100` I guess, if I'm not wrong.